File: packages/skillet/src/eval/providers/factory.ts
import type { ProviderConfig } from "../config/schema.js";
import { AnthropicProvider } from "./anthropic.js";
import { GoogleProvider } from "./google.js";
import { OpenAIProvider } from "./openai.js";
import type { LLMProvider } from "./types.js";

export function createProvider(config: ProviderConfig & { apiKey: string }): LLMProvider {
	switch (config.name) {
		case "anthropic":
			return new AnthropicProvider(config.apiKey, config.model);
		case "openai":
			return new OpenAIProvider(config.apiKey, config.model);
		case "google":
			return new GoogleProvider(config.apiKey, config.model);
		default:
			throw new Error(`Unknown provider: ${config.name}`);
	}
}

File: packages/skillet-eval-python/src/skill_eval/providers/factory.py
from __future__ import annotations

from skill_eval.config.schema import ResolvedGraderConfig, ResolvedProviderConfig
from skill_eval.providers.anthropic import AnthropicProvider
from skill_eval.providers.google import GoogleProvider
from skill_eval.providers.openai import OpenAIProvider
from skill_eval.providers.types import LLMProvider


def create_provider(
    config: ResolvedProviderConfig | ResolvedGraderConfig,
) -> LLMProvider:
    name = getattr(config, "name", None) or getattr(config, "provider", None)
    match name:
        case "anthropic":
            return AnthropicProvider(config.api_key, config.model)
        case "openai":
            return OpenAIProvider(config.api_key, config.model)
        case "google":
            return GoogleProvider(config.api_key, config.model)
        case _:
            raise ValueError(f"Unknown provider: {name}")
